src/mpi/coll/src/csel_container.c

```diff
@@ -550,5 +692,8 @@ void *MPII_Create_container(struct json_object *obj)
         MPL_free(ckey);
     }
 
+    /* process algorithm parameters */
+    parse_container_params(json_object_object_get(obj, key), cnt);
```

Comment:
I dont see how this can be done. This is what you are asking for I guess -
 ```
405         else if (!strcmp(ckey, "algorithm=MPIR_Ibcast_intra_gentran_tree")) {
 406             cnt->id = MPII_CSEL_CONTAINER_TYPE__ALGORITHM__MPIR_Ibcast_intra_gentran_tree;
 407             if (!strncmp(ckey, "chunk_size=", strlen("chunk_size=")))
 408                 cnt->u.ibcast.intra_gentran_tree.chunk_size = atoi(ckey + strlen("chunk_size="));
 409             else if (!strncmp(ckey, "tree_type=", strlen("tree_type=")))
 410                 cnt->u.ibcast.intra_gentran_tree.tree_type = atoi(ckey + strlen("tree_type=");
 411             else if (!strncmp(ckey, "k=", strlen("k=")))
 412                 cnt->u.ibcast.intra_gentran_tree.k = atoi(ckey + strlen("k="));
 413         }
```

The problem is both "algo" and "parameters" are in `ckey`. And not all parameters apply to all algorithms. So in the `container_parse_function`, we need to parse the `ckey` again to fill the containers created in `create_container()`

Reply:
No, I meant you have the `obj` in the previous `foreach` loop, so you can parse that object to the the remaining entries of the container.

Reply:
I see. I will do it

Reply:
I put more time in this, but I actually like the way we are doing it now. I dont see any problem in that.

Reply:
It looks cleaner than having a `foreach` loop for each of the algorithms, if I want to further dig in `obj`

Reply:
Alright, I guess that's just a preference issue and I I don't want to hold this up for such a minor thing.

Reply:
Thanks

Reply:
Sorry, one additional thing: I looked at the code again to see what exactly was bothering me about it.  I think it's the fact that the parsing is based on "if the parameter is `chunk_size`", then apply it to algorithm X, Y, Z.   I think it's more natural if the parsing was the other way around.  That is, first check for the algorithm and then find all the parameters that apply to it.  Can this be done?

Reply:
This is how I ideally want to do it - 
```
        else if (!strcmp(ckey, "algorithm=MPIR_Ibcast_gentran_tree")) {
            cnt->id = MPII_CSEL_CONTAINER_TYPE__ALGORITHM__MPIR_Ibcast_intra_gentran_tree;
            struct json_object * p_obj = json_object_object_get(obj, key);
            json_object_object_foreach(p_obj, p_key, p_val) {
                char *ckey = MPL_strdup_no_spaces(p_key);
                if (!strncmp(ckey, "k=", strlen("k="))) {
                    cnt->u.ibcast.intra_gentran_tree.k = atoi(ckey + strlen("k="));
                }
            }
```
But I see several compiler errors. I cannot have `p_key, p_val`. And if I use `key, val` it complains about shadowing a previous variable.

Reply:
No, that's not what I meant.  I meant you can still use your `parse_container_params` function.  But in that function, can you flip the search for container parameters to branch first on the container type.  Something like this:

```
if (cnt->id == MPII_CSEL_CONTAINER_TYPE__ALGORITHM__MPIR_Ibcast_intra_gentran_tree) {
    json_object_object_foreach(obj, key, val) {
        char *ckey = MPL_strdup_no_spaces(key);
        if (!strncmp(ckey, "chunk_size=", strlen("chunk_size="))) {
            cnt->u.ibcast.intra_gentran_tree.chunk_size = chunk_size;
        } else if (... some other parameter ...) {
            cnt->u.ibcast.intra_gentran_tree.<other_param> = <val>;
        }
    }
}
```

This way, all the parameters associated with a given algorithm are localized instead of being spread out.

Reply:
Done

Reply:
It does not work. See the compiler errors.

Reply:
You are missing braces in each case statement.